torch/csrc/distributed/c10d/Functional.cpp

```diff
@@ -0,0 +1,362 @@
+#include <torch/csrc/distributed/c10d/Functional.hpp>
+
+#include <shared_mutex>
+
+#include <ATen/ATen.h>
+#include <ATen/core/op_registration/op_registration.h>
+#include <c10/core/DispatchKey.h>
+#include <torch/csrc/autograd/function.h>
+#include <torch/csrc/distributed/c10d/ProcessGroup.hpp>
+
+namespace {
+
+// NOTE [Process Group Resolution]
+//
+// c10d_functional requires process group objects to be associated with a tag.
+// Using this tag, the collective ops can determine the appropriate process
+// group object to use. The mechanism that maintains the association must
+// accommodate both the one-rank-per-process and one-rank-per-thread scenarios.
+// Specifically, in one-rank-per-thread scenario, different thread should
+// associate with different process group objects based on the same tag, making
+// the resolution logic context-dependent. While using thread_local might
+// appear as a feasible solution, it falls short during backward passes. This
+// is because an autograd thread cannot be reliably linked to a specific rank.
```

Comment:
iirc in autograd there's a "driver thread" and worker threads to which really execute the backward nodes. We need this autograd thread id because we don't know which thread we are in during backward, as backward nodes itself might contain collective?

Reply:
Exactly. In GPU workloads, each autograd worker thread maps to a single device. In theory, we could resolve the rank via the device. However, this doesn't work in CPU workloads where work assignment is arbitrary. The approach in this PR works for both GPU and CPU workloads.